### PR TITLE
[Improve][Connector-V2] Add nonblank validation for required Slack options

### PR DESCRIPTION
**Pull request by 1328837476-hug**
### Purpose of this pull request

Related to #11007. Add declarative `Conditions.notBlank()` validation for the three existing required Slack options: `webhooks_url`, `oauth_token`, and `slack_channel`.

This change is limited to `SlackSinkFactory` and focused factory tests. It does not modify the client, writer, or the OAuth token mapping fix in #12171.

### Does this PR introduce _any_ user-facing change?

Yes. Previously, empty or whitespace-only strings passed factory validation. These values are now rejected during configuration validation. All three options remain required, and all nonblank strings remain accepted without additional URL-format or token-prefix constraints.

### How was this patch tested?

Factory tests use `ConfigValidator` without contacting Slack and cover:

- A valid configuration.
- Each required option missing individually.
- Each option set to an empty string, a space, a tab, a newline, or mixed whitespace.
- Each option set to an arbitrary nonblank string with surrounding spaces.
- Validation errors identifying the affected option.

Validated locally on Windows with JDK 17 and the Maven Wrapper using the equivalent `mvnw.cmd` commands:

```text
./mvnw -f seatunnel-connectors-v2/connector-slack/pom.xml spotless:apply
./mvnw -pl seatunnel-connectors-v2/connector-slack -am -Dtest=SlackFactoryTest -Dsurefire.failIfNoSpecifiedTests=false verify
git diff --check
```

A temporary Maven settings file pointed to Maven Central because the local mirror had a certificate mismatch. All 10 test invocations passed with no failures, errors, or skipped tests; the blank-value tests iterate over the inputs listed above. All nine selected reactor modules passed `verify`. Spotless and `git diff --check` passed. No remote Slack integration tests were run.

### Check list

- No new dependencies or binary packages.
- Scope follows the maintainer's factory-validation-and-tests guidance; the behavior change is described above. Documentation is unchanged.
- No new connector, plugin mapping, distribution configuration, or CI label changes.

### CODE REVIEW

**Comment by nzw921rx** [comment]
please use EN,thanks

File: seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackFactoryTest.java [hunk 1]
```diff
@@ -17,15 +17,68 @@
 
 package org.apache.seatunnel.connectors.seatunnel.slack;
```
Comment (utafrali) [inline seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackFactoryTest.java:19]: Package mismatch: `SlackFactoryTest` is declared in `org.apache.seatunnel.connectors.seatunnel.slack`, but `SlackSinkFactory` lives in `...slack.sink`. The test file should be moved to `src/test/java/.../slack/sink/SlackFactoryTest.java` and the package updated to `org.apache.seatunnel.connectors.seatunnel.slack.sink` to follow the project convention of mirroring source packages in tests.

File: seatunnel-connectors-v2/connector-slack/src/main/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackSinkFactory.java [hunk 2]
```diff
@@ -36,10 +38,9 @@ public String factoryIdentifier() {
     @Override
     public OptionRule optionRule() {
         return OptionRule.builder()
-                .required(
-                        SlackSinkOptions.WEBHOOKS_URL,
-                        SlackSinkOptions.OAUTH_TOKEN,
-                        SlackSinkOptions.SLACK_CHANNEL)
+                .required(SlackSinkOptions.WEBHOOKS_URL, notBlank(SlackSinkOptions.WEBHOOKS_URL))
```
Comment (utafrali) [inline seatunnel-connectors-v2/connector-slack/src/main/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackSinkFactory.java:41]: The semantics of `OptionRule.Builder.required(Option<?>, Condition<?>...)` need to be confirmed before this can merge. In SeaTunnel, the `Condition` arguments to `required()` typically encode *when* the option is required (i.e. conditional requirement), not additional validation constraints. If that is the case, `required(WEBHOOKS_URL, notBlank(WEBHOOKS_URL))` reads as "WEBHOOKS_URL is required only when WEBHOOKS_URL is not blank" — circular logic that would silently accept both absent and blank values.

If the intended meaning is "always required AND must not be blank", please cite or link to another connector in the codebase using this exact pattern so reviewers can confirm it. If the conditional-requirement semantics apply, the fix is to keep unconditional `required(WEBHOOKS_URL)` and add a separate constraint for the notBlank check.

**Review by utafrali — COMMENTED** [review]
The scope is well-contained and the test coverage for missing and blank values is thorough. The blocking question is whether `OptionRule.Builder.required(option, notBlank(option))` truly means "always required AND notBlank" or the more common SeaTunnel reading of "required only when notBlank" — the latter would be circular and the validation would be a no-op. There is also a pre-existing bug in `SlackClient.publishMessage()` where `.token(pluginConfig.get(SLACK_CHANNEL))` should be `.token(pluginConfig.get(OAUTH_TOKEN))` — out of scope here but should be tracked separately.

File: seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/SlackFactoryTest.java [hunk 3]
```diff
@@ -17,15 +17,68 @@
...
+    void missingRequiredOption(String key) {
+        Map<String, Object> options = validOptions();
+        options.remove(key);
+        assertInvalid(options, key);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"webhooks_url", "oauth_token", "slack_channel"})
+    void blankRequiredOption(String key) {
+        for (String value : new String[] {"", " ", "\t", "\n", " \t\r\n "}) {
```
Comment (utafrali) [inline seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/SlackFactoryTest.java]: The whitespace variants are iterated inside a `@ParameterizedTest`, so a failure on `"\t"` only reports the outer `key` parameter — making it hard to tell which variant broke. Consider flattening to a `@MethodSource` that yields `(key, blankValue)` pairs, giving each combination its own named test case and a clearer failure message.

File: seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackFactoryTest.java [hunk 4]
```diff
@@ -17,15 +17,68 @@
...
+    @ParameterizedTest
+    @ValueSource(strings = {"webhooks_url", "oauth_token", "slack_channel"})
+    void blankRequiredOption(String key) {
+        for (String value : new String[] {"", " ", "\t", "\n", " \t\r\n "}) {
+            Map<String, Object> options = validOptions();
+            options.put(key, value);
+            assertInvalid(options, key);
+        }
+    }
+
```
Comment (utafrali) [inline seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackFactoryTest.java:69]: The assertion `exception.getMessage().contains(key)` is tied to the internal message format of `OptionValidationException`. If that format changes, this test will fail for reasons unrelated to the validation logic. This is low-risk today but worth noting as a fragility point.

File: seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackFactoryTest.java (hunk 1 above)
Comment (1328837476-hug) [reply seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackFactoryTest.java:19]: Thanks for the suggestion. The test already used this package before this PR, so this is not a compilation issue, but mirroring the factory's package would make the layout clearer. I will move the test to the `slack.sink` package in a follow-up update.

File: seatunnel-connectors-v2/connector-slack/src/main/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackSinkFactory.java (hunk 2 above)
Comment (1328837476-hug) [reply seatunnel-connectors-v2/connector-slack/src/main/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackSinkFactory.java:41]: Thanks for raising this. I checked the implementation at this PR's dev baseline: this overload means "unconditionally required AND satisfies the value constraint". It adds the option to `AbsolutelyRequiredOptions` and adds the conditions separately to `valueConstraints`: [OptionRule.Builder.required](https://github.com/apache/seatunnel/blob/72acda57042efa3408681dc2c93663910b06b3e8/seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/util/OptionRule.java#L313-L323).

An existing connector uses the exact same pattern: [ActivemqSinkFactory](https://github.com/apache/seatunnel/blob/72acda57042efa3408681dc2c93663910b06b3e8/seatunnel-connectors-v2/connector-activemq/src/main/java/org/apache/seatunnel/connectors/seatunnel/activemq/sink/ActivemqSinkFactory.java#L54-L55).

The factory tests in this PR also exercise each option missing individually and each blank input through ConfigValidator, and assert that validation throws. These tests passed locally, so absent and blank values are not silently accepted. I will retain the current factory rules.

File: seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/SlackFactoryTest.java (hunk 3 above)
Comment (1328837476-hug) [reply seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/SlackFactoryTest.java]: Agreed. I will replace the inner loop with a `@MethodSource` covering each option/blank-value combination, with readable labels for the whitespace variants so failures identify the exact case. This test reorganization is not yet included in the current commit.

File: seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackFactoryTest.java (hunk 4 above)
Comment (1328837476-hug) [reply seatunnel-connectors-v2/connector-slack/src/test/java/org/apache/seatunnel/connectors/seatunnel/slack/sink/SlackFactoryTest.java:69]: Thanks for noting this. The assertion only checks that the affected option's key appears in the error, not the complete message, punctuation, or wording. This verifies that a configuration error identifies the option the user needs to fix. I would prefer to retain this focused diagnostic assertion while avoiding an exact message match.

**Comment by 1328837476-hug** [comment]
@nzw921rx The PR title and description have been updated to English. Apologies for the earlier language mismatch.

@utafrali Thanks for the review. I have replied to the inline comments with the framework implementation and an existing connector example confirming the validation semantics. I will follow up with the test package and parameterization improvements. The OAuth token mapping issue is already tracked separately in #12171 and remains outside this PR.

**Comment by nzw921rx** [comment]
please enable CI 

image

**Comment by 1328837476-hug** [comment]
Updated in a25dcc986. Moved SlackFactoryTest to the slack.sink package and replaced the whitespace loop with individually named @MethodSource cases. All 22 factory tests and the existing client test passed; verify succeeded for all nine selected modules. The factory validation rules remain unchanged. Please take another look. Thanks!

**Review by DanielLeens — COMMENTED** [review]
# What Problem Does This PR Solve?
- **User pain point**: Before this change, `SlackSinkFactory#optionRule()` only checked that `webhooks_url`, `oauth_token`, and `slack_channel` were *present* in the config. A config with `webhooks_url: ""` (or all-whitespace) passed validation, and the user only found out something was wrong later, from a confusing runtime failure inside the Slack client rather than a clear config error at job submission time.
- **Fix approach**: Attach `Conditions.notBlank(...)` as a value constraint to each of the three required options via `OptionRule.Builder.required(Option, Condition...)`, so `ConfigValidator` now rejects blank/whitespace-only values with a clear `OptionValidationException` that names the offending key.
- **One-sentence summary**: A small, well-scoped connector-config-validation improvement that closes a real gap (blank values silently passing `required` checks) with solid test coverage, and the semantic question raised earlier in review has been correctly resolved with codebase evidence.

Welcome, and thanks for the contribution — congratulations on your first PR to Apache SeaTunnel! This is a clean, well-targeted change and the back-and-forth with @utafrali in the thread is exactly the kind of rigor we like to see. I did a fresh, full review of the current head (`a25dcc9`) below.

# 1. Code Change Review

## 1.1 Core Logic Analysis

**Change**: `seatunnel-connectors-v2/connector-slack/src/main/java/.../sink/SlackSinkFactory.java:41-43`

```java
// before
.required(
        SlackSinkOptions.WEBHOOKS_URL,
        SlackSinkOptions.OAUTH_TOKEN,
        SlackSinkOptions.SLACK_CHANNEL)

// after
.required(SlackSinkOptions.WEBHOOKS_URL, notBlank(SlackSinkOptions.WEBHOOKS_URL))
.required(SlackSinkOptions.OAUTH_TOKEN, notBlank(SlackSinkOptions.OAUTH_TOKEN))
.required(SlackSinkOptions.SLACK_CHANNEL, notBlank(SlackSinkOptions.SLACK_CHANNEL))
```

I independently traced the semantics that @utafrali questioned, rather than taking the author's reply at face value, since this is exactly the kind of thing that's worth verifying from source before merge:

- `OptionRule.Builder.required(Option<?> option, Condition<?> condition1, Condition<?>... conditions)` (`seatunnel-api/.../OptionRule.java:313-324`) adds the option to `AbsolutelyRequiredOptions` (unconditional presence requirement) **and separately** appends the condition to `valueConstraints`. These are two independent checks, not "required only if the condition holds" — the overload is correctly named/documented as "Absolutely required option with value constraints" and the class-level Javadoc example at the top of the file (`OptionRule.java:34`, `.required(HOST, notBlank(HOST))`) is exactly this pattern.
- I confirmed `ConfigValidator.collectErrors` (`ConfigValidator.java:190-249`) runs the presence check (`checkRequiredOption`) and the `valueConstraints` check independently; a structurally-absent key is skipped in the constraint loop only to avoid a duplicate/misleading error for something already reported as missing (`ConfigValidator.java:230-233`), not because presence implies "constraint satisfied."
- I confirmed the actual gap this PR closes: `ConfigValidator.hasOption()` (`ConfigValidator.java:411-413`) is `config.getOptional(option).isPresent()` — an explicit `""` value is "present," so without `notBlank` a blank string sails through the old `required(...)` check untouched. This is a real, reproducible bug fix, not a cosmetic change.
- I also independently verified the cited precedent, `ActivemqSinkFactory` (`seatunnel-connectors-v2/connector-activemq/.../ActivemqSinkFactory.java:52-53`), uses the identical `.required(QUEUE_NAME, notBlank(QUEUE_NAME))` pattern. So this is consistent with an established codebase convention, not a novel/risky usage.

**Conclusion**: @utafrali's semantic concern was a legitimate thing to check pre-merge, and the author's reply fully and correctly resolves it with primary-source evidence (source line ranges + a working precedent). I verified both independently and they check out. This is a normal-path change — every job submission that builds this sink factory's `OptionRule` and runs it through `ConfigValidator` (i.e. every Slack sink job) now gets the stricter check, so this isn't a narrow edge-case fix, it applies on the main config-validation path.

**Scope check**: `git diff --stat` against `dev` touches exactly the sink factory and its test (plus a test file move) — no other file in `connector-slack` is touched. Slack has no source factory (sink-only connector), so there's no missing sibling factory that should have received the same treatment. The pre-existing bug @utafrali flagged in `SlackClient.publishMessage()` (`.token(pluginConfig.get(SLACK_CHANNEL))` should be `OAUTH_TOKEN`) is genuinely out of scope here — I confirmed it was already fixed and merged separately in #12171 ("[Fix][Connector-V2] Use OAuth token for Slack messages"), so no action needed in this PR.

## 1.2 Compatibility Impact

**Partially incompatible, by design and acceptable**: Any existing job configuration that supplied a blank/whitespace-only value for `webhooks_url`, `oauth_token`, or `slack_channel` will now fail fast at config-validation time instead of failing later (or silently misbehaving) at runtime. That is the intended effect of the fix. A genuinely valid configuration (non-blank values) is completely unaffected — validated by the `preserveNonblankValues` test. I consider this an acceptable, desirable compatibility change: it converts a latent runtime failure into an explicit, actionable config error, which is strictly better for users. No other behavior, field, or default value changes.

## 1.3 Performance / Side-Effect Analysis

None. This is a pure validation-time change (string blank-check on three fields at job submission), with no impact on the runtime write path, no new allocations of consequence, and no change to `SlackClient`'s behavior.

## 1.4 Error Handling and Logging

No High or Medium severity issues found here. One Low item:

- **Issue 1** — Location: `seatunnel-connectors-v2/connector-slack/src/test/java/.../sink/SlackFactoryTest.java:81`. Problem: `assertInvalid` asserts `exception.getMessage().contains(key)`, which is coupled to `OptionValidationException`'s internal message formatting. Potential risk: if the message format changes elsewhere in `seatunnel-api`, this test could fail for reasons unrelated to the Slack connector's validation logic. Best improvement: none required right now — this was already raised by @utafrali, and the author's reasoning (verifying the message names the broken key, not exact wording) is reasonable and low-risk. Severity: Low, non-blocking.

# 2. Code Quality Assessment

## 2.1 Coding Standards

Import ordering (static import for `notBlank` placed after regular/third-party imports, matching `ActivemqSinkFactory`'s convention) is correct. The change itself is a 3-line declarative config-rule tweak with no new methods/fields, consistent in style and (lack of) inline commentary with every other `*SinkFactory#optionRule()` in this codebase (e.g. `ActivemqSinkFactory`) — I don't think additional comments are warranted here; the code reads clearly as-is and matching the established, comment-free convention for this kind of declarative rule is the right call.

## 2.2 Test Coverage and Test Stability

Coverage is thorough: `validConfig`, `missingRequiredOption` (per-key absence), `blankRequiredOption` (per-key × 5 blank variants: empty/space/tab/newline/mixed whitespace), and `preserveNonblankValues` (regression guard that a legitimately non-blank value, including one with surrounding whitespace, is still accepted). That directly covers the "absent" and "present-but-blank" boundary this PR is meant to fix.

**Stability rating: Stable.** Checked against the flaky-test anti-pattern checklist:
- No `Thread.sleep` / async waits — this is pure synchronous validation logic, no async involved.
- No nondeterministic assertions — inputs and expected outcomes are fully deterministic (fixed strings/whitespace variants).
- No shared mutable static state — `validOptions()` builds a fresh `HashMap` per test invocation (`SlackFactoryTest.java` in the `sink` package); no cross-test leakage.
- No order dependence — each `@Test`/`@ParameterizedTest` is self-contained and does not rely on execution order.
- No weak E2E readiness checks — this is a unit test hitting `ConfigValidator` directly, not an integration/E2E harness.

The prior round's feedback on this front (package mismatch and the flattened `@MethodSource` with named whitespace variants, per-`(key, blankValue)` case) is fully addressed in the current head: the test now lives in `.../slack/sink/SlackFactoryTest.java` (`sink` package, mirroring `SlackSinkFactory`'s package, confirmed at line 18 of the file) and `blankRequiredOptions()` (`SlackFactoryTest.java:58-67`) yields named `(key, description, value)` triples via `@ParameterizedTest(name = "{0}: {1}")`, so a failure on e.g. the `"\t"` case for `oauth_token` will now show up as its own distinctly-named test case instead of being buried inside an outer loop. Good responsiveness to review feedback.

## 2.3 Documentation Updates

None needed. `docs/en/connectors/sink/Slack.md` already lists all three options as `required: Yes`; this PR only tightens enforcement of an already-documented requirement, it doesn't change the option contract, so the docs remain accurate as-is.

# 3. Architectural Soundness

## 3.1 Elegance of the Solution

**Precise fix.** This uses the framework's existing, purpose-built mechanism (`OptionRule.Builder.required(Option, Condition...)` + `Conditions.notBlank`) exactly as it's intended and precedented elsewhere (`ActivemqSinkFactory`), rather than hand-rolling ad-hoc blank checks inside the factory or the client. This is the smallest, most idiomatic change that fully closes the gap.

## 3.2 Maintainability

High. Declarative, self-documenting, and consistent with the rest of the codebase's `optionRule()` conventions — a future maintainer skimming this file alongside any other connector factory will immediately recognize the pattern.

## 3.3 Extensibility

N/A — this is a leaf-level connector config rule, not a shared/public abstraction, so there's no broader extensibility surface to reason about. Nothing here touches a shared class or common entry point.

## 3.4 Historical-Version Compatibility

No checkpoint/state/serialization surface is touched (Slack is a fire-and-forget message sink, no split/checkpoint state). No upgrade or restore path is affected. The only compatibility consideration is the config-validation behavior change already covered in 1.2, which is a job-submission-time behavior change only, not a data/state-format change.

# 4. Issue Summary

| Number | Issue | Location | Severity |
| --- | --- | --- | --- |
| 1 | `assertInvalid` couples the test to `OptionValidationException`'s message format via `contains(key)` | `SlackFactoryTest.java:81` | Low |

# 5. Merge Recommendation

### Conclusion: Ready to merge after fixes

1. Blockers — none. I found no functional, correctness, or compatibility blockers in the current head.
2. Recommended fixes — non-blocking:
   - Issue 1 above (test assertion coupling) — optional, author already made a reasonable case for keeping it.

The one item that actually needs to happen before this can be merged isn't code: the `Build` check is currently stuck in `queued` status with no corresponding workflow run recorded yet (`gh api .../actions/runs` shows no `Build`/backend workflow run for this head SHA), which lines up with @nzw921rx's "please enable CI" comment and screenshot — this looks like a first-time-contributor workflow-approval gate rather than an actual CI failure. A maintainer with write access will need to approve running the workflow from the Actions tab so `Build` can actually execute; I don't have that permission from this account. Once CI has actually run and is green, I don't see anything else standing in the way.

Overall assessment: this is a well-scoped, well-tested fix for a real validation gap, and the author's handling of @utafrali's review — citing exact source line ranges and a working precedent instead of just asserting "trust me" — is exactly the kind of evidence-based response that makes review efficient. No alternative implementation is needed; the chosen approach (Option A, using the existing `required(Option, Condition...)` + `notBlank` framework mechanism) is already the right one — there's no meaningfully different Option B here that would be better. Nice work, and thanks for iterating quickly on the earlier feedback.

**Comment by DanielLeens** [comment]
CI update: the workflow-approval gate mentioned above is cleared. The `Build` check on `a25dcc9` was stuck in `queued` with no recorded run when I last looked; it has since actually run (started 23:52 UTC, completed ~01:11 UTC) and finished green across the matrix, alongside `Notify test workflow` and `labeler`. No other blocker remains from my side.

Congrats again on your first PR making it all the way through CI, and thanks for the careful back-and-forth on the validation semantics earlier in the thread — this is ready for a committer to take a final look and merge whenever they have a moment.

**Comment by 1328837476-hug** [comment]
Thanks for checking the CI results and for the thorough review/

**Review by nzw921rx — APPROVED** [review]
+1 LGTM
